druid/src/widget/aspect_ratio_box.rs

```diff
@@ -0,0 +1,164 @@
+    /// will become 1.
+    pub fn set_ratio(&mut self, ratio: f64) {
+        self.ratio = clamp_ratio(ratio);
+    }
+
+    /// Generate `BoxConstraints` that fit within the provided `BoxConstraints`.
+    ///
+    /// If the generated constraints do not fit then they are constrained to the
+    /// provided `BoxConstraints`.
+    fn generate_constraints(&self, bc: &BoxConstraints) -> BoxConstraints {
+        let (mut new_width, mut new_height) = (bc.max().width, bc.max().height);
+
+        if new_width == f64::INFINITY {
+            new_width = new_height * self.ratio;
+        } else {
+            new_height = new_width / self.ratio;
+        }
+
+        if new_width > bc.max().width {
+            new_width = bc.max().width;
+            new_height = new_width / self.ratio;
+        }
+
+        if new_height > bc.max().height {
+            new_height = bc.max().height;
+            new_width = new_height * self.ratio;
+        }
+
+        if new_width < bc.min().width {
+            new_width = bc.min().width;
+            new_height = new_width / self.ratio;
+        }
+
+        if new_height < bc.min().height {
+            new_height = bc.min().height;
+            new_width = new_height * self.ratio;
+        }
+
+        BoxConstraints::tight(bc.constrain(Size::new(new_width, new_height)))
+    }
+}
+
+/// Clamps the ratio between 0.0 and f64::MAX
+/// If ratio is 0.0 then it will return 1.0 to avoid creating NaN
+fn clamp_ratio(mut ratio: f64) -> f64 {
+    ratio = f64::clamp(ratio, 0.0, f64::MAX);
+
+    if ratio == 0.0 {
+        warn!("Provided ratio was <= 0.0.");
+        1.0
+    } else {
+        ratio
+    }
+}
+
+impl<T: Data> Widget<T> for AspectRatioBox<T> {
+    #[instrument(
+        name = "AspectRatioBox",
+        level = "trace",
+        skip(self, ctx, event, data, env)
+    )]
+    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut T, env: &Env) {
+        self.inner.event(ctx, event, data, env);
+    }
+
+    #[instrument(
+        name = "AspectRatioBox",
+        level = "trace",
+        skip(self, ctx, event, data, env)
+    )]
+    fn lifecycle(&mut self, ctx: &mut LifeCycleCtx, event: &LifeCycle, data: &T, env: &Env) {
+        self.inner.lifecycle(ctx, event, data, env)
+    }
+
+    #[instrument(
+        name = "AspectRatioBox",
+        level = "trace",
+        skip(self, ctx, old_data, data, env)
+    )]
+    fn update(&mut self, ctx: &mut UpdateCtx, old_data: &T, data: &T, env: &Env) {
+        self.inner.update(ctx, old_data, data, env);
+    }
+
+    #[instrument(
+        name = "AspectRatioBox",
+        level = "trace",
+        skip(self, ctx, bc, data, env)
+    )]
+    fn layout(&mut self, ctx: &mut LayoutCtx, bc: &BoxConstraints, data: &T, env: &Env) -> Size {
+        bc.debug_check("AspectRatioBox");
+
+        if bc.max() == bc.min() {
+            warn!("Box constraints are tight. Aspect ratio box will not be able to preserve aspect ratio.");
```

Comment:
The conditions for this warning are not complete. I think the layout algorithm will fail if `bc.min().height * ratio < bc.max().width`, or if `bc.min().height / ratio < bc.max().width`. Could have my math wrong there but it's something similar.

That said, I'm not convinced this always represents invalid state. Aside from precedent (`SizedBox` doesn't warn if it can't meet its constraints), I think there's an argument here for allowing "messy" box constraints. What if I sometimes want to control the width and height and sometimes want to let the constraints work? I'd at least want a way to turn off the warning.

Reply:
I agree with you here, that there are other situations where the widget can't meet the desired aspect ratio. I wouldn't be opposed to warning in those cases, but I think it would be kinda noisy.
Conversely, if the BoxConstraints are tight, and don't match the desired ratio already, this widget is not permitted to solve that problem... And there are only a few ways to get tight bcs - usually, it's because of a SizedBox setting the size, or because you're the root widget of the window (or any direct descendant before splitting or padding has been added).
I don't have a strong opinion here about whether the warning stays, but I do see the logic in warning about this situation.

Reply:
Hmm, I see what you're saying. I can still imagine a warning in a "valid" case where the box constraints are not constant, but it's probably enough of an edge case to not matter too much? If a warning or two can be considered okay in a valid but far-out edge case then I'm not worried about it.

Reply:
Right, I think this could show up spuriously in some edge cases, for sure. And I think a broader "I can't hit the desired ratio" warning would come up a lot more in spurious or transient situations... but I think either or both warnings could be good.
I'm not sure what the right signal-noise ratio is here, but we've usually tended toward only high-snr warnings.

Reply:
I think maybe what I can do is instead of warning there, I can run the generate constraints, then check if the final size maintains the aspect ratio. If it doesn't then I can warn and print the box constraints it was given and the aspect ratio?

Reply:
That way I'm not checking for all the conditions where it could fail, though that may be less helpful to the user?